File: 0001~0500/0069. Sqrt(x)/main.cpp
/** 
 * Source: https://leetcode.com/problems/sqrtx/
 * Date: 2021/12/31
 * Skill: Binary Search
 * Runtime: 
 * Memory Usage: 
 * Time complexity:
 * Space complexity: 
 * Constraints: 
 *      0 <= x <= 2^31 - 1
 *     
 */

#include<iostream>

using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        if (x == 0 || x == 1) return x;
        else return mySqrt2(0, x, x);
    }

    int mySqrt2(int l, int h, int x) {
        if (l == h) return l;
        int m = (l + h) / 2;
        if (m * m == x) return m;
        else if (m * m > x) return mySqrt2(l, m, x);
        else return mySqrt2(m, h, x);
    }
};

int main() {
    /*Solution s;
    cout << s.mySqrt(8) << endl;
    return 0;*/
    int a = 1;
    for (int i = 1; i < 9; i++) {
        a = a * i;
        cout << a << endl;
    }
}

File: 0001~0500/0069. Sqrt(x)/main.py
"""
Source: https://rb.gy/hjjxb
Date: 2023/7/24
Skill:
Ref:
Runtime: 51 ms, faster than 72.58%
Memory Usage: 16.42 MB, less than 27.29%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def mySqrt(self, x: int) -> int:
        lo, hi = 0, x
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if mid * mid == x:
                return mid
            elif mid * mid < x:
                lo = mid
            else:
                hi = mid - 1

        return lo


if __name__ == "__main__":
    s = Solution()
